### DISCUSSION TITLE
Vercel deployed Server Component state not working well

### DISCUSSION THREAD

**Discussion by Gyurmatag** [pmndrs/zustand#2122] [Q&A]
Hey!
I've got a little demo site here deployed on Vercel: https://weather-app-lac-eta-80.vercel.app/
Here is the Github repo: https://github.com/Gyurmatag/weather-app

My problem is that when I add a new city to the store like this in a Server component...
```
  async function saveCity(data: FormData) {
    'use server';

    const newCityKey = String(Object.keys(useStore.getState()).length);

    useStore.setState((state) => ({
      ...state,
      [newCityKey]: {
        name: data.get('cityName') as string,
        latitude: data.get('latitude') as string,
        longitude: data.get('longitude') as string,
      },
    }));

    redirect('/');
  }
```

It is working well in local environment but not in my Vercel deploy. Somehow it doesn't get added to the store.
Can you please help me what would be the problem?

Thank you very much!

**Comment by dai-shi** [comment]
I don't think it should work in the local env.
For example, if you open two browser windows, you will notice the store value is shared across session, which should be undesirable.

It should be reproducible without zustand, with a module variable instead.

**Reply by Gyurmatag** [reply]
Yes, that is exactly what I want. Here is an article about it: https://blog.stackademic.com/minimalistic-guide-zustand-next-js-server-components-22ebcb00eb8f
It works fine in local, but not on Vercel

**Reply by dai-shi** [reply]
Sorry if my comment was ambiguous.

Client state management like Zustand simply doesn't work on server, because server doesn't have state.
So, use `useStore` always under the client tree.

**Reply by Gyurmatag** [reply]
In this video it shows that it can store the state on the server:
https://youtu.be/OpMAH2hzKi8?si=_YZgDrIJc1MIIvfw&t=426
But okay, I will make that component 'client side' so it will not be a problem...

**Reply by dai-shi** [reply]
I don't think we know the best practice yet, but it will probably be using it with `cache`. Yeah, meanwhile, using it on client side would be good.

**Reply by nicitaacom** [reply]
> but it will probably be using it with `cache`. Yeah, meanwhile, using it on client side would be good.

Exactly - I would like to store products in user cart on a server to keep SSR in next js

**Reply by Gyurmatag** [reply]
@nicitaacom You can use databasea to store product in user cart. In that way the user can access the cart in another device too.

**Reply by nicitaacom** [reply]
> You can use databasea

I know - and I do it
I want to save products in cart for not authenticated user - create table in PostgreSQL table and clear this table not my case - but it is an option
I would like to store it in cookies rather than local storage - but as I know it inpossible with zustand unfortuneately
